source/common/upstream/logical_dns_cluster.cc

```diff
@@ -87,7 +99,9 @@ void LogicalDnsCluster::startResolve() {
             current_resolved_address_ = new_address;
             // Capture URL to avoid a race with another update.
             tls_->runOnAllThreads([this, new_address]() -> void {
-              tls_->getTyped<PerThreadCurrentHostData>().current_resolved_address_ = new_address;
+              PerThreadCurrentHostData& data = tls_->getTyped<PerThreadCurrentHostData>();
+              data.current_resolved_address_ = new_address;
+              data.context_ = context_;
```

Comment:
Technically, you should capture context as a lambda variable and assign it. It's not safe to access cross thread.